src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -415,15 +502,231 @@ private void CheckDisposedOrStarted()
+            }
+        }
+
+        private SafeCurlSlistHandle SetRequestHeaders(SafeCurlHandle handle, HttpRequestMessage request)
+        {
+            SafeCurlSlistHandle retVal = new SafeCurlSlistHandle();
+            if (request.Headers == null)
+            {
+                return retVal;
+            }
+
+            HttpHeaders contentHeaders = null;
+            if (request.Content != null)
+            {
+                // TODO: Content-Length header isn't getting correctly placed using ToString()
+                // This is a bug in HttpContentHeaders that needs to be fixed.
+                if (request.Content.Headers.ContentLength.HasValue)
+                {
+                    long contentLength = request.Content.Headers.ContentLength.Value;
+                    request.Content.Headers.ContentLength = null;
+                    request.Content.Headers.ContentLength = contentLength;
+                }
+                contentHeaders = request.Content.Headers;
+            }
+
+            string[] allHeaders = HeaderUtilities.DumpHeaders(request.Headers, contentHeaders)
+                .Split(s_headerDelimiters, StringSplitOptions.RemoveEmptyEntries);
+            bool gotReference = false;
+            try
+            {
+                for (int i = 0; i < allHeaders.Length; i++)
+                {
+                    string header = allHeaders[i];
+                    if (header.Equals("{") || header.Equals("}"))
+                    {
+                        continue;
+                    }
+                    IntPtr rawHandle = IntPtr.Zero;
+                    if (!retVal.IsInvalid)
+                    {
+                        retVal.DangerousAddRef(ref gotReference);
+                        rawHandle = retVal.DangerousGetHandle();
+                    }
+                    rawHandle = Interop.libcurl.curl_slist_append(rawHandle, header);
+                    retVal.SetHandle(rawHandle);
+                }
+            }
+            finally
+            {
+                if (gotReference)
+                {
+                    retVal.DangerousRelease();
+                }
+            }
+
+            if (!retVal.IsInvalid)
+            {
+                try
+                {
+                    retVal.DangerousAddRef(ref gotReference);
+                    SetCurlOption(handle, CURLoption.CURLOPT_HTTPHEADER, retVal.DangerousGetHandle());
+                }
+                finally
+                {
+                    if (gotReference)
+                    {
+                        retVal.DangerousRelease();
+                    }
+                }
+            }
+
+            return retVal;
+        }
+
+        private void AddEasyHandle(RequestCompletionSource state)
+        {
+            bool gotReference = false;
+            SafeCurlHandle requestHandle = state.RequestHandle;
+            try
+            {
+                requestHandle.DangerousAddRef(ref gotReference);
```

Comment:
I think it'd be worth commenting that you indeed intend to not be releasing this in the successful case, because you're giving out the pointer to native code for the duration of the operation, and you'll be releasing it later.